File: backend/agents/workerSafetyAgent.js
// backend/agents/workerSafetyAgent.js
// Agent 2 — Worker Safety Agent
// Checks PPE, risky positions, fall hazards, equipment proximity

import { callGemmaVision } from './gemmaClient.js';

const WORKER_SAFETY_PROMPT = `You are the Worker Safety Agent for ConstructGuard AI.

You specialize in OSHA construction safety standards (29 CFR 1926).

Look at this construction site image and identify ALL worker safety issues.

Respond ONLY with valid JSON:
{
  "violations": [
    {
      "worker_id": "W-01",
      "issue": "No hard hat",
      "osha_standard": "29 CFR 1926.100",
      "severity": "CRITICAL|HIGH|MEDIUM|LOW",
      "location": "description of where in the frame"
    }
  ],
  "ppe_compliance": {
    "hard_hat": "compliant|non-compliant|partial|unknown",
    "safety_vest": "compliant|non-compliant|partial|unknown",
    "safety_boots": "compliant|non-compliant|partial|unknown",
    "gloves": "compliant|non-compliant|partial|unknown",
    "eye_protection": "compliant|non-compliant|partial|unknown"
  },
  "fall_hazards": ["list any fall hazard situations"],
  "equipment_proximity_risks": ["list workers too close to moving machinery"],
  "blocked_exits": ["list any blocked emergency exits"],
  "overall_compliance_score": 0-100,
  "summary": "One sentence summary"
}`;

export async function runWorkerSafetyAgent({ imageBase64, imageMimeType }) {
  console.log('[WorkerSafetyAgent] Checking PPE and hazards...');
  if (!imageBase64) return { violations: [], ppe_compliance: {}, fall_hazards: [], overall_compliance_score: 100, summary: 'No image provided' };
  return callGemmaVision(imageBase64, imageMimeType, WORKER_SAFETY_PROMPT);
}

File: backend_py/agents/worker_safety_agent.py
from typing import Any

from models.gemma_adapter import gemma

WORKER_SAFETY_PROMPT = '''You are the Worker Safety Agent for ConstructGuard AI.
You specialize in OSHA construction safety standards, especially 29 CFR 1926.
Look at this construction site image and identify worker safety issues.
Respond only with valid JSON:
{
  "violations": [],
  "ppe_compliance": {
    "hard_hat": "compliant|non-compliant|partial|unknown",
    "safety_vest": "compliant|non-compliant|partial|unknown",
    "safety_boots": "compliant|non-compliant|partial|unknown",
    "gloves": "compliant|non-compliant|partial|unknown",
    "eye_protection": "compliant|non-compliant|partial|unknown"
  },
  "fall_hazards": [],
  "equipment_proximity_risks": [],
  "blocked_exits": [],
  "overall_compliance_score": 0,
  "summary": "one sentence"
}'''


def fallback(message: str = 'Local vision model unavailable') -> dict[str, Any]:
    return {
        'violations': [],
        'ppe_compliance': {'hard_hat': 'unknown', 'safety_vest': 'unknown', 'safety_boots': 'unknown', 'gloves': 'unknown', 'eye_protection': 'unknown'},
        'fall_hazards': [],
        'equipment_proximity_risks': [],
        'blocked_exits': [],
        'overall_compliance_score': 0,
        'summary': message,
    }


def run_worker_safety_agent(image_base64: str | None, image_mime_type: str | None) -> dict[str, Any]:
    if not image_base64:
        return fallback('No image provided')
    try:
        return gemma.analyze_image(image_base64, image_mime_type or 'image/jpeg', WORKER_SAFETY_PROMPT)
    except Exception as exc:
        data = fallback('Worker safety agent used fallback output')
        data['error'] = str(exc)
        return data
